man/man8/timelord.8
'\" t
.\"     Title: timelord
.\"    Author: [FIXME: author] [see http://www.docbook.org/tdg5/en/html/author]
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 15 Feb 2023
.\"    Manual: Netatalk 2.3.0
.\"    Source: Netatalk 2.3.0
.\"  Language: English
.\"
.TH "TIMELORD" "8" "15 Feb 2023" "Netatalk 2.3.0" "Netatalk 2.3.0"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
timelord \- Macintosh time server daemon
.SH "SYNOPSIS"
.HP \w'\fBtimelord\fR\ 'u
\fBtimelord\fR [\-d] [\-l] [\-n\ \fInbpname\fR]
.SH "DESCRIPTION"
.PP
\fBtimelord\fR
is a simple time server for Macintosh computers that use the
\fBtardis\fR
client\&. It has the same capabilities as the "Timelord" CDEV for Macintosh\&.
.PP
\fBtardis\fR
is implemented as a Chooser extension\&. In the Chooser, choose the
\fBtimelord \fR
instance to synchronize the Mac\*(Aqs system time with\&. Once synchronized,
\fBtardis\fR
will automatically synchronize with the server at boot, or scheduled at regular intervals while the Mac is running (the latter requires tardis 1\&.4)\&.
.SH "OPTIONS"
.PP
\fB\-d\fR
.RS 4
Debug mode, i\&.e\&. don\*(Aqt disassociate from controlling TTY\&.
.RE
.PP
\fB\-l\fR
.RS 4
Return the time zone adjusted localtime of the server\&. The default behavior without this option is GMT\&.
.RE
.PP
\fB\-n\fR \fInbpname\fR
.RS 4
Register this server as
\fInbpname\fR\&. This defaults to the hostname\&.
.RE


NEWS
Changes in 2.3.0
================
* afpd: Remove cdb and tdb CNID backends, GitHub #535
* afpd: Remove nostat option, GitHub #540
* afpd: Remove the "Force UID/GID" feature, GitHub #516
* afpd: Remove experimental Dropbox Kludge feature, GitHub #515
* afpd: Remove afprun module from codebase, GitHub #526
* afpd: Deprecate AFS (Andrew File System) support, GitHub #554
* afpd: Enable only DHX2 UAM by default, GitHub #358
* afpd: Validate size of ace_count in FPSetACL request, GitHub #364
* afpd: Enable, sort, and flesh out Mac file type translation, GitHub #534
* afpd: Don't seteuid() if process is already running as that uid, GitHub #532
* afpd: Use correct username length in afp_changepw, GitHub #627
* uams: Link PGP UAM with CRYPT_LIBS, GitHub #548
* uams: Use correct data type to build with PAM support on FreeBSD, GitHub #563
* atalkd: Introduce 3rd party bridge quirks mode for Asante and Dayna bridges, GitHub #585
* atalkd: Revert registering of NBP entries of "Workstation" and "netatalk", GitHub #473
* papd: Use packed structs for pap packets, GitHub #527
* papd: Use a portable fcntl() call instead of flock(), GitHub #625
* timelord: Flip the check for tm->tm_gmtoff to appease FreeBSD, GitHub #415
* config: Name the default home dir shared volume in AppleVolumes.default, GitHub #469
* distrib: Add Documentation tags to systemd templates, GitHub #378
* distrib: Write a Description and Short-Description for the Debian init script, GitHub #428
* docs: Remove redundant AUTHORS file, GitHub #454
* docs: Add README for the v2.3 branch, GitHub #588
* docs: Remove obsoleted bug reporting sections, GitHub #455
* docs: Remove redundant id properties, GitHub #520
* docs: Use absolute XSL stylesheet path, source VERSION in manpages script, GitHub #458
* docs: Update URLs in website, GitHub #445
  - Point to the root of SF downloads where new files are added
  - Load CSS from the current protocol and domain
* docs: Standardize man page heading case, GitHub #545
* docs: Minor cleanup of man pages, GitHub #470
* docs: Document the correct default log level for afpd, GitHub #552
* docs: Improve readability of manual README, GitHub #440
* docs: Document binheader, nadheader, and showpap, GitHub #430
* docs: Fix user visible typos and improve English grammar, GitHub #381 #382
* docs: Create man page for cnid2_create tool, GitHub #402
* docs: Fix typos, improve layout in afpd.conf man page, GitHub #472
* build system: Default to xz compression for tarballs, GitHub #483
* build system: Remove vestiges of a Webmin install target, GitHub #467
* build system: Fold a2boot and timelord under the appletalk conditional, GitHub #423
* build system: Don't install at.h when glibc header found, GitHub #409
* build system: Switch from compile-time debug to run-time debug, GitHub #536
* build system: Correct install hook for static config files, GitHub #408
* build system: Make libdir the default install location for UAM libs, GitHub #407
* build system: Option to skip privileged hooks for make install, GitHub #405
* build system: Rearrange the dummy paths for the distribution tests, GitHub #522
* build system: Configurable systemd prefix, with default to /usr/lib, GitHub #404
  - Remove hard-coded systemd path /lib/systemd in Makefile
  - Introduce check for the two known common prefixes
  - Introduce --with-systemd-prefix autoconf option
* build system: Fix detection of Berkeley DB installed in multiarch location, GitHub #380
* build system: Support building against libtirpc as separate from glibc, GitHub #385
* build system: Recommend BerkeleyDB 5.3, GitHub #8
* Port 2.x fork code quality improvements GitHub #488
  - Remove support for ancient OSes:
    SunOS, IRIX, AIX, Ultrix, HPUX, Tru64, GNU/kFreeBSD, UnixWare
  - Remove ancient bug workarounds and standard library shortcomings
  - Refactor functions
  - More granular typecasting
  - Initialize variables
  - Prevent potential memory leaks
  - Handle system call return values
* Improve logger_test, while re-enabling the syslog tests, GitHub #584
* Rename asip-status.pl to asip-status, GitHub #379
* Clean up residual CVS commit headers, GitHub #631

Changes in 2.2.10
=================
* UPD: The ability to build and run on macOS Ventura, GH#350
       - Removed legacy conditionals for Mac OS X Server
       - Modernized autotools syntax
* UPD: papd: CUPS API improvements, GH#371
       - Fix printing to Avahi discovered "driverless" printers
       - Remove CUPS PPD API dependency, which is slated for deprecation
       - Allow papd to return a printer resolution with LaserWriter 8
* UPD: Remove redundant documentation in conf files, GH#333
       - Add man page paragraph on the illegalseq option
* FIX: Use non-interactive PAM session to avoid spurious error, GH#361
* FIX: Adopt Debian downstream patches, GH#369
       - Drop bogus warning in AppleVolumes.default 
       - Fix syntax and typo in add_netatalk_printer script
       - Fix user-visible typos in log output and documentation
       - Improvements to Debian legacy init.d script
* FIX: atalkd.service register NBP entries "Workstation" and "netatalk", GH#372
* FIX: Restore tarball distribution of READMEs in doc/ GH#374

Changes in 2.2.9
================

* FIX: Fix printing from old Mac OS LaserWriter drivers
* FIX: CVE-2022-45188
* FIX: Improve systemd service dependencies GH#232
* FIX: macusers: Fix output for long usernames
* FIX: macusers: account for usernames with non-word characters
* FIX: macusers: do not show root user. SF bug #495
* NEW: Add manpage for a2boot
* UPD: Manpage and html manual updates

Changes in 2.2.8
================

* NEW: asip-status.pl: IPv6 support; show GSS-UAM SPNEGO blob;
       improved layout of output. (3.1 backport)
* NEW: apple_dump: support for EA meta data. (3.1 backport)
* NEW: Import netatalk-doc into the main repo, and overhaul
       scripts, man pages and html manual sources.
* UPD: Display the Netatalk Daemon icon with the '-icon' afpd.conf

contrib/timelord/timelord.c
/*
 * $Id: timelord.c,v 1.8 2005-04-28 20:49:36 bfernhomberg Exp $
 *
 * Copyright (c) 1990,1992 Regents of The University of Michigan.
 * All Rights Reserved. See COPYRIGHT.
 *
 * The "timelord protocol" was reverse engineered from Timelord,
 * distributed with CAP, Copyright (c) 1990, The University of
 * Melbourne.  The following copyright, supplied by The University
 * of Melbourne, may apply to this code:
 *
 *	This version of timelord.c is based on code distributed
 *	by the University of Melbourne as part of the CAP package.
 *
 *	The tardis/Timelord package for Macintosh/CAP is
 *	Copyright (c) 1990, The University of Melbourne.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

#include <sys/param.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/file.h>
#include <sys/uio.h>
#include <stdlib.h>

#include <unistd.h>

#include <netatalk/at.h>
#include <netatalk/endian.h>
#include <atalk/atp.h>
#include <atalk/nbp.h>

#include <time.h>
#ifdef HAVE_SGTTY_H
#include <sgtty.h>
#endif /* HAVE_SGTTY_H */
#include <errno.h>
#include <signal.h>
#include <atalk/logger.h>
#include <stdio.h>
#include <string.h>
#ifdef HAVE_NETDB_H
#include <netdb.h>
#endif /* HAVE_NETDB_H */

#ifdef HAVE_FCNTL_H
#include <fcntl.h>
#endif /* HAVE_FCNTL_H */
#ifdef HAVE_SYS_FCNTL_H
#include <fcntl.h>
#endif /* HAVE_SYS_FCNTL_H */

#include <termios.h>
#include <sys/ioctl.h>

#define	TL_GETTIME	0
#define	TL_OK		12
#define TL_BAD		10
#define EPOCH		0x7C25B080	/* 00:00:00 GMT Jan 1, 1970 for Mac */
#define SIZEOF_MAC_LONG 4

int	debug = 0;
char	*bad = "Bad request!";
char	buf[ 4624 ];
char	*server;

void usage( char *p )
{
    char	*s;

    if (( s = rindex( p, '/' )) == NULL ) {
	s = p;
    } else {
	s++;
    }
    fprintf( stderr, "Usage:\t%s -d -l -n nbpname\n", s );
    exit( 1 );
}

/*
 * Unregister ourself on signal.
 */
void
goaway(int signal)
{
    if ( nbp_unrgstr( server, "TimeLord", "*", NULL ) < 0 ) {
	LOG(log_error, logtype_default, "Can't unregister %s", server );
	exit( 1 );
    }
    LOG(log_info, logtype_default, "going down" );
    exit( 0 );
}

int main( int ac, char **av )
{
    ATP			atp;
    struct sockaddr_at	sat;
    struct atp_block	atpb;
    struct timeval	tv;
    struct iovec	iov;
    struct tm		*tm;
    char		hostname[ MAXHOSTNAMELEN ];
    char		*p;
    int			c;
    int         lflag = 0;
    long		req, mtime, resp;
    extern char		*optarg;
    extern int		optind;

    if ( gethostname( hostname, sizeof( hostname )) < 0 ) {
	perror( "gethostname" );
	exit( 1 );
    }
    if (( server = index( hostname, '.' )) != 0 ) {
	*server = '\0';
    }
    server = hostname;

    while (( c = getopt( ac, av, "dln:" )) != EOF ) {
	switch ( c ) {
	case 'd' :
	    debug++;
	    break;
    case 'l':
        lflag = 1;
        break;
	case 'n' :
	    server = optarg;
	    break;
	default :
	    fprintf( stderr, "Unknown option -- '%c'\n", c );
	    usage( *av );
	}
    }

    /*
     * Disassociate from controlling tty.
     */
    if ( !debug ) {
	int		i, dt;

	switch ( fork()) {
	case 0 :
	    dt = getdtablesize();
	    for ( i = 0; i < dt; i++ ) {
		(void)close( i );
	    }
	    if (( i = open( "/dev/tty", O_RDWR )) >= 0 ) {
		(void)ioctl( i, TIOCNOTTY, 0 );
		setpgid( 0, getpid());
		(void)close( i );
	    }
	    break;
	case -1 :
	    perror( "fork" );
	    exit( 1 );
	default :
	    exit( 0 );
	}
    }

    if (( p = rindex( *av, '/' )) == NULL ) {
	p = *av;
    } else {
	p++;
    }

    set_processname(p);
    syslog_setup(log_debug, logtype_default, logoption_ndelay|logoption_pid, logfacility_daemon );

    /* allocate memory */
    memset (&sat.sat_addr, 0, sizeof (sat.sat_addr));

    if (( atp = atp_open( ATADDR_ANYPORT, &sat.sat_addr )) == NULL ) {
	LOG(log_error, logtype_default, "main: atp_open: %s", strerror( errno ) );
	exit( 1 );
    }

    if ( nbp_rgstr( atp_sockaddr( atp ), server, "TimeLord", "*" ) < 0 ) {
	LOG(log_error, logtype_default, "Can't register %s", server );
	exit( 1 );
    }
    LOG(log_info, logtype_default, "%s:TimeLord started", server );
    if ( lflag == 1 ) {
        LOG(log_info, logtype_default, "Time zone is localtime" );
    }
    else {
        LOG(log_info, logtype_default, "Time zone is GMT" );
    }

	signal(SIGHUP, goaway);
	signal(SIGTERM, goaway);

    for (;;) {
	/*
	 * Something seriously wrong with atp, since these assigns must
	 * be in the loop...
	 */
	atpb.atp_saddr = &sat;
	atpb.atp_rreqdata = buf;
	bzero( &sat, sizeof( struct sockaddr_at ));
	atpb.atp_rreqdlen = sizeof( buf );
	if ( atp_rreq( atp, &atpb ) < 0 ) {
	    LOG(log_error, logtype_default, "main: atp_rreq: %s", strerror( errno ) );
	    exit( 1 );
	}

	p = buf;
	bcopy( p, &req, SIZEOF_MAC_LONG);
	req = ntohl( req );
	p += SIZEOF_MAC_LONG;

	switch( req ) {
	case TL_GETTIME :
	    if ( atpb.atp_rreqdlen > 5 ) {
		    bcopy( p + 1, &mtime, SIZEOF_MAC_LONG);
		    mtime = ntohl( mtime );
		    LOG(log_info, logtype_default, "gettime from %s %s was %lu",
			    (*( p + 5 ) == '\0' ) ? "<unknown>" : p + 5,
			    ( *p == 0 ) ? "at boot" : "in chooser",
			    mtime );
	    } else {
		    LOG(log_info, logtype_default, "gettime" );
	    }

	    if ( gettimeofday( &tv, NULL ) < 0 ) {
		LOG(log_error, logtype_default, "main: gettimeofday: %s", strerror( errno ) );
		exit( 1 );
	    }

        if ( lflag == 1 ) {
            if (( tm = localtime( &tv.tv_sec )) == 0 ) {
            perror( "localtime" );
            exit( 1 );
            }

#if defined (NO_STRUCT_TM_GMTOFF)
	mtime = tv.tv_sec + EPOCH - timezone;
#else /* NO_STRUCT_TM_GMTOFF */
	mtime = tv.tv_sec + EPOCH + tm->tm_gmtoff;
#endif /* NO_STRUCT_TM_GMTOFF */
        }
        else {
            if (( tm = gmtime( &tv.tv_sec )) == 0 ) {
            perror( "gmtime" );
            exit( 1 );
            }

            mtime = tv.tv_sec + EPOCH;
        }

	    mtime = htonl( mtime );

	    resp = htonl( TL_OK );
	    bcopy( &resp, buf, SIZEOF_MAC_LONG);
	    bcopy( &mtime, buf + SIZEOF_MAC_LONG, SIZEOF_MAC_LONG);
	    iov.iov_len = SIZEOF_MAC_LONG + SIZEOF_MAC_LONG;
	    break;

	default :
	    LOG(log_error, logtype_default, bad );

	    resp = htonl( TL_BAD );
	    bcopy( &resp, buf, SIZEOF_MAC_LONG);
	    *( buf + 4 ) = (unsigned char)strlen( bad );
	    strcpy( buf + 5, bad );
	    iov.iov_len = SIZEOF_MAC_LONG + 2 + strlen( bad );
	    break;
	}

	iov.iov_base = buf;
	atpb.atp_sresiov = &iov;
	atpb.atp_sresiovcnt = 1;
	if ( atp_sresp( atp, &atpb ) < 0 ) {
	    LOG(log_error, logtype_default, "main: atp_sresp: %s", strerror( errno ) );
	    exit( 1 );
	}
    }
}
